crates/uv/tests/pip_compile.rs

```diff
@@ -3910,3 +3940,68 @@ fn no_deps_invalid_extra() -> Result<()> {
 
     Ok(())
 }
+
+/// Resolve a package from a `requirements.in` file, with a `constraints.txt` file pinning one of
+/// its transitive dependencies to a specific version.
+#[test]
+fn compile_constraints_compatible_version() -> Result<()> {
+    let context = TestContext::new("3.12");
+    let requirements_in = context.temp_dir.child("requirements.in");
+    requirements_in.write_str("virtualenv")?;
+
+    let constraints_txt = context.temp_dir.child("constraints.txt");
+    constraints_txt.write_str("filelock==3.8.0")?;
+
+    uv_snapshot!(context.compile()
+            .arg("requirements.in")
+            .arg("--constraint")
+            .arg("constraints.txt"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    # This file was autogenerated by uv via the following command:
+    #    uv pip compile --cache-dir [CACHE_DIR] --exclude-newer 2023-11-18T12:00:00Z requirements.in --constraint constraints.txt
+    distlib==0.3.7
+        # via virtualenv
+    filelock==3.8.0
+        # via virtualenv
+    platformdirs==3.11.0
+        # via virtualenv
+    virtualenv==20.21.1
+
+    ----- stderr -----
+    Resolved 4 packages in [TIME]
+    "###
+    );
+
+    Ok(())
+}
+
+/// Resolve a package from a `requirements.in` file, with a `constraints.txt` file pinning one of
+/// its direct dependencies to an incompatible version.
+#[test]
+fn compile_constraints_incompatible_version() -> Result<()> {
+    let context = TestContext::new("3.12");
+    let requirements_in = context.temp_dir.child("requirements.in");
+    requirements_in.write_str("filelock==1.0.0")?;
+
+    let constraints_txt = context.temp_dir.child("constraints.txt");
+    constraints_txt.write_str("filelock==3.8.0")?;
+
+    uv_snapshot!(context.compile()
+            .arg("requirements.in")
+            .arg("--constraint")
+            .arg("constraints.txt"), @r###"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+
+    ----- stderr -----
+      × No solution found when resolving dependencies:
+      ╰─▶ Because you require filelock==1.0.0 and you require filelock==3.8.0, we
+          can conclude that the requirements are unsatisfiable.
```

Comment:
This one definitely makes me want to open an issue to display the locations of direct requirements in error messages.